pkgs/applications/science/math/mxnet/default.nix

```diff
@@ -2,11 +2,10 @@
 , opencv3, gtest, blas, gomp, llvmPackages, perl
 , cudaSupport ? config.cudaSupport or false, cudaPackages ? {}, nvidia_x11
 , cudnnSupport ? cudaSupport
-, cudaCapabilities ? [ "3.7" "5.0" "6.0" "7.0" "7.5" "8.0" "8.6" ]
```

Comment:
This change removes user's ability to customize `cudaCapabilities` on a per-package basis. Standardizing `cudaCapabilities` across nixpkgs is a big step forward, but it cannot come at the cost of losing per-package customization.

What about something like
```nix
, cudaCapabilities ? cudaFlags.cudaCapabilities
```
instead?

Reply:
At first glance I liked this idea, but I'm struggling with the right way to implement it.

- `cudaCapabilities` was inconsistent used ("5.0" vs "sm50" vs "compute50" vs "Maxwell").  So if I restore the original arguments then the overrides may also be different on a per-library basis.  That feels confusing and undesirable as an interface. Some of these even had their own local translation logic.
- If I implement a consistent `cudaCapabilities` to all the libraries, I would need to locally use the mappers to translate them into the variants need for the builder tools.  I prefer this route, but it is less clean.

But from a more practical standpoint, I cannot think of many real-world scenarios where you want your Tensorflow to be compiled for a different set of Cuda architectures than your Pytorch.  Is it really something we expect people to do?  If so, would an intermediate (for `overrideAttr`) be enough?

Reply:
> > If I implement a consistent cudaCapabilities to all the libraries, I would need to locally use the mappers to translate them into the variants need for the builder tools. I prefer this route, but it is less clean.
> 
> Yes, this is what I was thinking. I'm assuming that this kind of translation must be done on a per-package basis anyhow? Thus is the price of normalization?

We generate all the mappings based on the hardware spec (user choice, SASS/PTX, optional forward compatibility) in a single place then just choose the correct args at the library level.

So code like:
```
builtins.concatStringsSep "," cudaFlags.cudaRealArchs
```

expands to:
```
  cudaCapabilitiesString = lib.strings.concatStringsSep ","
    (cudaFlags.cudaArchMapper "sm"
      (if cudaCapabilities != null then cudaCapabilities else cudaFlags.cudaCapabilities));
```

Which works for changing the hardware generations, but won't deal with more complex cases like mixing SASS and PTX.  Some packages even optionally take in cudaPackages, so you cannot just default it.

> The use case I had in mind was for someone working with a cloud VM who wants to use one VM/GPU spec for some software and a different VM/GPU spec for other software. In a desktop situation, a user might have two GPUs of different models. But on second thought, I concede that these are relatively unique scenarios and workarounds are still possible. Happy to proceed here however you feel is most straightforward

I'm open to modifications to the PR if you have a suggested approach (Nix is not my day-to-day, so I could be missing a good idiom here), but if you no longer feel strongly this is something people will use then I'd propose keeping things clean and readable for the majority of users.

Users will always have the ability to look at the generated flag like `cMakeFlags` and just override as needed or create multiple platforms at the top-level (a more consistent approach given the ML library interdependences here).

Reply:
> Some packages even optionally take in cudaPackages, so you cannot just default it.

Yes, this is the pattern that we are in the process of migrating towards atm.

> I'm open to modifications to the PR if you have a suggested approach (Nix is not my day-to-day, so I could be missing a good idiom here), but if you no longer feel strongly this is something people will use then I'd propose keeping things clean and readable for the majority of users.

Thanks for your thorough explanation. I understand better now. Personally I think I would have opted for a single, standardized definition of capabilities in `cudaFlags` and then helper functions to map those flags into common representations, with each package putting the global flags together with the appropriate serialization. But I'm really splitting hairs here. I think this looks solid, and it will be a welcome change relative to the current SOTA

Marking this as resolved